[sw/silicon_creator] Add flash_ctrl driver

jon-flatley opened the pull request:
Add a flash_ctrl driver implementation.
This driver supports:

- flash_ctrl initialization.
- Read transactions via flash_ctrl read FIFO.
- Status queries (read FIFO full/empty, init status, transaction status, error status.)

Keeping this as WIP for now as it lacks unit tests and complete error handling.

### Open Questions

- [ ] What sort of error handling do we want in mask_rom? [flash_ctrl error codes](https://docs.opentitan.org/hw/ip/flash_ctrl/doc/index.html#Reg_err_code) for reference.
- [ ] Do we need program/erase functionality also in mask_rom?

tjaychen commented:
just a quick note, i mentioned this to @cfrantz also.  Some of the flash features (specifically around memory attributes for protection, scrambling, ecc), are not very well verified yet.  So I would recommend only lightly using them for now, or if we you see weird bugs, roll that back and see if it helps.

General page program / erase / read SHOULD work okay, since otherwise I would have expected the FPGAs to have blown up by now...

cfrantz commented on sw/device/silicon_creator/lib/drivers/flash_ctrl.c:
```diff
@@ -0,0 +1,100 @@
+          control_reg_val, FLASH_CTRL_CONTROL_INFO_SEL_FIELD, 2u);
+      break;
+    default:
+      return kErrorUnknown;
```
Create an error value specifically for this case.

I thought I had documented this, but it isn't in errors.h: The value `kErrorUnknown` should be used by any modules.  It is a value that all errors will be converted to at the error reporting stage _if_ we're configured for maximum error redaction.

bbhanushali replied:
@jon-flatley PTAL

cfrantz commented on sw/device/silicon_creator/lib/drivers/flash_ctrl.c:
```diff
@@ -0,0 +1,100 @@
+                      bitfield_bit32_write(0u, FLASH_CTRL_INIT_VAL_BIT, true));
+}
+
+flash_ctrl_status_t flash_ctrl_status_get(const flash_ctrl_t *flash_ctrl);
```
Does this have an implementation?  I can't seem to find it.

bbhanushali replied:
@jon-flatley PTAL

cfrantz commented on sw/device/silicon_creator/lib/drivers/flash_ctrl.c:
```diff
@@ -0,0 +1,100 @@
+      control_reg_val, FLASH_CTRL_CONTROL_PARTITION_SEL_BIT, 1u);
+
+  // Set the partition and info type.
+  switch (region) {
+    case kFlashCtrlRegionData:
+      // 0 indicates data partition.
+      control_reg_val = bitfield_bit32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_PARTITION_SEL_BIT, 0u);
+      break;
+    case kFlashCtrlRegionInfo0:
+      // The partition is already set to info, so just set the info type.
+      control_reg_val = bitfield_field32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_INFO_SEL_FIELD, 0u);
+      break;
+    case kFlashCtrlRegionInfo1:
+      control_reg_val = bitfield_field32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_INFO_SEL_FIELD, 1u);
+      break;
+    case kFlashCtrlRegionInfo2:
+      control_reg_val = bitfield_field32_write(
+          control_reg_val, FLASH_CTRL_CONTROL_INFO_SEL_FIELD, 2u);
```
This is ok for now, but we should consider if there is a way to be clever with the values of flash_control_region_select_t so as to avoid having to have this switch/case translator.

IOW, could we write this bit of code as something like:
```
  control_reg_val |= (uint32_t)region
```

bbhanushali replied:
@jon-flatley PTAL

jon-flatley commented:
Apologies for the slow turnaround on this. Latest push addresses most comments and add unit tests.

> This is ok for now, but we should consider if there is a way to be clever with the values of flash_control_region_select_t so as to avoid having to have this switch/case translator.

@cfrantz I haven't included this, but if you still think this is a good idea I'm happy to adopt this. I personally don't like assigning magic values to enums like this as imo it adds unnecessary pitfalls down the road if they ever need changing.

cfrantz commented on sw/device/silicon_creator/lib/drivers/flash_ctrl.c:
```diff
@@ -0,0 +1,126 @@
+  return kErrorOk;
+}
+
+int flash_ctrl_fifo_read(int word_count, uint32_t *data_out) {
```
Should `word_count` be of type `size_t`?
Should the return value be `size_t` as well?

jon-flatley replied:
Yeah probably, fixed

alphan commented:
> What sort of error handling do we want in mask_rom? flash_ctrl error codes for reference.

I think any type of flash error in mask ROM is unexpected and should probably lead to shutdown. Error codes, especially ECC-related ones probably need more consideration but `flash_ctrl_status_get` doesn't read those.

> Do we need program/erase functionality also in mask_rom?

ROM_EXT: yes, mask ROM: yes, (for bootstrap, @cfrantz) but I don't think this should block this PR.

cfrantz commented:
> > What sort of error handling do we want in mask_rom? flash_ctrl error codes for reference.
> 
> I think any type of flash error in mask ROM is unexpected and should probably lead to shutdown. Error codes, especially ECC-related ones probably need more consideration but `flash_ctrl_status_get` doesn't read those.

Generally, I agree with this, but I think "lead to shutdown" can be at the discretion of the caller of any of the `flash_ctrl` function.  Basically, the `RETURN_IF_ERROR` construction will in most cases lead back to the "main" function and exit into `shutdown_finalize`.

> 
> > Do we need program/erase functionality also in mask_rom?
> 
> ROM_EXT: yes, mask ROM: yes, (for bootstrap, @cfrantz) but I don't think this should block this PR.

Yes, we need program and erase functionality in the mask_rom specifically for bootstrap.  These functions can be added as a follow-up PR.

jon-flatley commented:
- Updated the FIFO read to rely on back pressure for control flow and removed impacted tests
- Updated `flash_ctrl_partition_t` to have the same bit pattern as what is written to the control register
- Addressed some nits

A followup PR will be needed to add program/erase functionality and to implement the correct init behavior.

cfrantz commented on sw/device/silicon_creator/lib/drivers/flash_ctrl.h:
```diff
@@ -0,0 +1,123 @@
+} flash_ctrl_region_select_t;
+
+/**
+ * Start a program transaction.
```
+1: s/program/read/

alphan reviewed (approved):
Thanks @jon-flatley! Could you please review the unresolved comments? It looks like most of them are already addressed but this [typo](https://github.com/lowRISC/opentitan/pull/6402#discussion_r660806121) still needs to be fixed.
